sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -595,32 +605,106 @@ size_t coap_put_option_uri(uint8_t *buf, uint16_t lastonum, const char *uri, uin
+        if (NANOCOAP_BLOCK_SZX_MAX - 4 < szx ) {
+            szx =NANOCOAP_BLOCK_SZX_MAX - 4;
+        }
+    }
+    blk->opt = buf;
+    blk->start = blknum * coap_szx2size(szx);
+    blk->end = blk->start + coap_szx2size(szx);
+    blk->cur = 0;
+    return coap_put_option_block2(buf, lastonum, blknum, szx, 1);
+}
+
+ssize_t coap_block2_build_reply(coap_pkt_t *pkt, unsigned code,
+                        uint8_t *rbuf, unsigned rlen, unsigned payload_len,
+                        coap_blockhelper_t *blk)
+{
+    if (blk->cur < blk->start) {
+        return coap_build_reply(pkt, COAP_CODE_BAD_OPTION, rbuf, rlen, 0);
+    }
+    int option_len;
+    uint16_t delta;
+    uint8_t *data_start = _parse_option(pkt, blk->opt, &delta, &option_len);
+    uint32_t blkopt = _decode_uint(data_start, option_len);
+    uint32_t blknum = blkopt >> COAP_BLOCKWISE_NUM_OFF;
+    unsigned szx = blkopt & COAP_BLOCKWISE_SZX_MASK;
+    int more = (blk->cur > blk->end) ? 0x80 : 0;
+
+    coap_put_option_block2(blk->opt, COAP_OPT_BLOCK2 - delta, blknum, szx, more);
+    return coap_build_reply(pkt, code, rbuf, rlen, payload_len);
+}
+
+size_t coap_blockwise_put_char(coap_blockhelper_t *blk, uint8_t *bufpos, char c)
+{
+    /* Only copy the char if it is within the window */
+    if (blk->start <=  blk->cur && blk->cur < blk->end) {
+        *bufpos = c;
+        blk->cur++;
+        return 1;
+    }
+    blk->cur++;
+    return 0;
+}
+
+size_t coap_blockwise_put_bytes(coap_blockhelper_t *blk, uint8_t *bufpos,
+                                 const uint8_t *c, size_t len)
+{
+    size_t str_len = 0;    /* Length of the string to copy */
+
+    /* Calculate start offset of the supplied string */
+    size_t str_offset = (blk->start > blk->cur) ? blk->start - blk->cur : 0;
+
+    /* Check for string before or beyond window */
+    if (blk->cur >= blk->end || str_offset > len) {
+        blk->cur += len;
+        return 0;
+    }
+    /* Check if string is over the end of the window */
+    if (blk->cur + len >= blk->end) {
+        str_len = blk->end - blk->cur - str_offset;
+    }
+    else {
+        str_len = len - str_offset;
+    }
+
+    /* Only copy the relevant part of the string to the buffer */
+    memcpy(bufpos, c + str_offset, str_len);
+    blk->cur += len;
+    return str_len;
+}
+
 ssize_t coap_well_known_core_default_handler(coap_pkt_t *pkt, uint8_t *buf, \
                                              size_t len, void *context)
 {
     (void)context;
-
+    coap_blockhelper_t blk;
     uint8_t *payload = buf + coap_get_total_hdr_len(pkt);
-
     uint8_t *bufpos = payload;
 
     bufpos += coap_put_option_ct(bufpos, 0, COAP_CT_LINK_FORMAT);
+    bufpos += coap_block2_init(bufpos, COAP_OPT_CONTENT_FORMAT, pkt, &blk);
     *bufpos++ = 0xff;
 
     for (unsigned i = 0; i < coap_resources_numof; i++) {
         if (i) {
-            *bufpos++ = ',';
+            bufpos += coap_blockwise_put_char(&blk, bufpos, ',');
```

Comment:
Does it make sense to have a single function that handles both a block write as well as a plain payload write (like *bufpos++ )? This would help avoid buffer overruns in both situations. Handles a plain payload if the blk pointer is NULL. How about calling it 'coap_payload_put_\[char|bytes\]()?

Reply:
Agree to resolve later as you look into it further.